scapy/fields.py

```diff
@@ -2128,3 +2128,52 @@ def i2repr(self, pkt, x):
         elif self.use_msec:
             x = x / 1e3
         return "%s sec" % x
+
+
+class ScalingField(Field):
+    __slots__ = ["scaling", "unit", "offset", "ndigits"]
+
+    def __init__(self, name, default, scaling=1.0, unit="",
+                 offset=0.0, ndigits=3, fmt="B"):
+        self.scaling = scaling
+        self.unit = unit
+        self.offset = offset
+        self.ndigits = ndigits
+        Field.__init__(self, name, default, fmt)
+
+    def i2m(self, pkt, x):
+        if x is None:
+            x = 0
+        return round((x - self.offset) / self.scaling)
+
+    def m2i(self, pkt, x):
+        return round(x * self.scaling + self.offset, self.ndigits)
+
+    def any2i(self, pkt, x):
+        if isinstance(x, int) and x >= 0:
+            x = struct.unpack(self.fmt,
+                              struct.pack(self.fmt.upper(), x))[0]
+            x = self.m2i(pkt, x)
+        return x
+
+    def i2repr(self, pkt, x):
+        return "%s %s" % (self.i2h(pkt, x), self.unit)
+
+
+class ScalingIntField(ScalingField):
+    # Avoid roundings to prevent precision loss
+    def __init__(self, name, default, scaling=1, unit="",
+                 offset=0, fmt="B"):
```

Comment:
The name `ScalingIntField` is easy to misunderstand. All `IntFields` in Scapy dissect usually 4 bytes. This isn't the case for this Field. Can you give me an example, where you lose the precision through rounding? If you want to prevent `round` from producing fractional digits, you can use the parameter `ndigits=0`.

Reply:
I agree with you. The name is misleading.

The problem is not the round, but the conversion to `float` in general.
Example:
```
>>> 18446744073709551615 == 18446744073709551615.0
False
```
This happens for very high numbers (long). OBD_PID7F has three of those long fields (8 bytes).
This is also a potential solution:
`ScalingField('total', 0, unit='sec', fmt='Q', offset=0, ndigits=0, scaling=1),`
Since it sets the default float values to integer values and thus prevents converting to float. But it feels like a hacky workaround.

Suggestions?

Reply:
I pushed an commit

Reply:
Does not work, because:
```
>>> 5 - 3.0
2.0
```

And `offset` is a float by default. `offset` has to be `0` (no float!) too.
Maybe it's the best to set the default values from `offset` to `0` and for `scaling` to `1`, instead of requiring to pass the integer value to the constructor.

Reply:
Ok, I'll think about the current implementation, and may refactor it in the next days

Reply:
Another example to think about. What behaviour would you expect? A or B?
```
>>> class DebugPacket(Packet):
...:    fields_desc = [
...:        ScalingField('data', 0, scaling=10)
...:    ]
...:
...:x = DebugPacket()
...:
...:x.data = 0x01
...:print(x.data)
A: 10.0
B: 1
>>> bytes(x)
A: '\x01'
B: '\x00'       => because of rounding
>>> x.data = 1.0
>>> print(x.data)
A: 1.0
B: 1
>>> bytes(x)
A: '\x00'       => because of rounding
B: '\x00'       => because of rounding
>>> x.data = 10
>>> print(x.data)
A: 100.0
B: 10
>>> bytes(x)
A: '\x0A'
B: '\x01' 
>>> x.data = 10.0
>>> print(x.data)
A: 10.0
B: 10
>>> bytes(x)
A: '\x01'
B: '\x01'      
>>> x.data = 0x0f
>>> print(x.data)
A: 150.0
B: 15
>>> bytes(x)
A: '\x0f'
B: '\x02'    => because of rounding
```

Reply:
At the moment, the `ScalingField` implements behaviour A. If we want to have Integers as internal storage, we should consider an implementation which behaves like B.

Reply:
That's a very good question.
I feel like the best option is to work always with the "raw" values, and only use the scaling in the `show` methods. Otherwise this would be very confusing. I actually needed a few minutes before I understood the output of A in the first example. Thus I would expect B in the most examples you provided.

Only for the most, because:

```python
>>> x.data = 10
>>> print(x.data)
A: 100.0
B: 10
>>> bytes(x)
A: '\x0A'
B: '\x01' 
```

In the first one I would expect B. But in the second one I would expect A, no matter if `data` is a float or an integer.


In summary, my idea is: The field contains the **data**, the `show` methods display the **information**. `bytes(x)` shall contain the same value as in the properties (like `data`).
`ScalingField` might provide a method which provides the implementation of A, thus converting the actual information to the necessary data.

Does this make sense to you and is it even implementable?

Reply:
Your summary makes sense. There are advantages and disadvantages of both strategies (using float vs. raw values internally). In my opinion, the **information value** is more important than the **raw value**. I think, that is also the present philosophy of Scapy. A big disadvantage when using raw values internally is the assignment of information values. We would need to create a complete new method for the conversion of an information value to a raw value, in order to be able to assign that value. 

Here is a list of features, we have to consider for ScalingFields:

- Assignment of information values `x.data = 150.0'
- Assignment of raw values `x.data = 0x0f` or `x.data = 15`
- Representation of information values with proper units
- Handling of signed information values
- Handling of information values with offset

Reply:
I'm pretty happy with the latest implementation. 
Last open question. Should we use a different name for this field. I'm not so happy with ScalingField. 
@Maintainers: How should we name this field? Maybe ScaledField or UnitField or ValueField, or something completely different.